File: ruby/longest_path.rb

def is_valid(x,y)
    if x < 0 or y < 0 or x >= N or y >= N 
        return false
    end

    return true
end

def longest_path(mat, visited, current_long, incoming_long, x, y)
    if x == 5 and y == 7 
        if incoming_long > current_long 
            current_long = incoming_long
            return current_long
        end
    end

    visited[x][y] = true
    i = 0 
    while i < 4 
        newX = x + ROW[i]
        newY = y + COL[i]
        if is_valid(newX, newY) and visited[newX][newY] == false and mat[newX][newY] == 1
            current_long = longest_path(mat, visited, current_long, incoming_long + 1, newX, newY)
        end
        i += 1
    end
    visited[x][y] = false

    return current_long
end

N = 10
mat = Array.new(N){Array.new(N)}
mat = [ 
    [ 1, 0, 1, 1, 1, 1, 0, 1, 1, 1 ],
    [ 1, 0, 1, 0, 1, 1, 1, 0, 1, 1 ],
    [ 1, 1, 1, 0, 1, 1, 0, 1, 0, 1 ],
    [ 0, 0, 0, 0, 1, 0, 0, 1, 0, 0 ],
    [ 1, 0, 0, 0, 1, 1, 1, 1, 1, 1 ],
    [ 1, 1, 1, 1, 1, 1, 1, 1, 1, 0 ],
    [ 1, 0, 0, 0, 1, 0, 0, 1, 0, 1 ],
    [ 1, 0, 1, 1, 1, 1, 0, 0, 1, 1 ],
    [ 1, 1, 0, 0, 1, 0, 0, 0, 0, 1 ],
    [ 1, 0, 1, 1, 1, 1, 0, 1, 0, 0 ]
]
visited = Array.new(N){Array.new(N)}
i = 0
while i < N 
    j = 0
    while j < N 
        visited[i][j] = false
        j += 1
    end
    i += 1
end

ROW = [-1,0,1,0]
COL = [0,1,0,-1]
incoming_long = 0
current_long = 0
max_dist = longest_path(mat, visited, current_long, incoming_long, 0, 0)
puts max_dist

File: cpp/longest_path.cpp
#include <iostream>
using namespace std;

const int N = 10;

const int ROW[4] = {-1, 0, 1, 0};
const int COL[4] = {0, 1, 0, -1};

int mat[N][N] = {
    {1, 0, 1, 1, 1, 1, 0, 1, 1, 1},
	{1, 0, 1, 0, 1, 1, 1, 0, 1, 1},
	{1, 1, 1, 0, 1, 1, 0, 1, 0, 1},
	{0, 0, 0, 0, 1, 0, 0, 1, 0, 0},
	{1, 0, 0, 0, 1, 1, 1, 1, 1, 1},
	{1, 1, 1, 1, 1, 1, 1, 1, 1, 0},
	{1, 0, 0, 0, 1, 0, 0, 1, 0, 1},
	{1, 0, 1, 1, 1, 1, 0, 0, 1, 1},
	{1, 1, 0, 0, 1, 0, 0, 0, 0, 1},
	{1, 0, 1, 1, 1, 1, 0, 1, 0, 0}
};

bool isValid(int x, int y) {
    if(x < 0 || y < 0 || x >= N || y >= N) {
        return false;
    }

    return true;
}

int findLongestPath(int mat[N][N], bool visited[N][N], int x, int y, int dist, int maxDist) {
    cout << x << y << endl;
    if(x == 5 && y == 7) {
        return max(dist, maxDist);
    }
    visited[x][y] = true;
    for(int i = 0; i < 4; i++) {
        int newX = x + ROW[i];
        int newY = y + COL[i];
        if(isValid(newX, newY) && mat[newX][newY] == 1 && visited[newX][newY] == false) {
            maxDist = findLongestPath(mat, visited, newX, newY, dist + 1, maxDist);
        }
    }
    visited[x][y] = false;

    return maxDist;
}

int main(){
    bool visited[N][N] = {false};
    int maxDist = findLongestPath(mat,visited, 0, 0, 0, 0);
    cout << maxDist << endl;
    return 0;
}




